source/windows/bcrypt_aes.c

```diff
@@ -0,0 +1,1110 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include <aws/cal/private/symmetric_cipher_priv.h>
+
+#include <windows.h>
+
+/* keep the space to prevent formatters from reordering this with the Windows.h header. */
+#include <bcrypt.h>
+
+#define NT_SUCCESS(status) ((NTSTATUS)status >= 0)
+
+/* handles for AES modes and algorithms we'll be using. These are initialized once and allowed to leak. */
+static aws_thread_once s_aes_thread_once = AWS_THREAD_ONCE_STATIC_INIT;
+static BCRYPT_ALG_HANDLE s_aes_cbc_algorithm_handle = NULL;
+static BCRYPT_ALG_HANDLE s_aes_gcm_algorithm_handle = NULL;
+static BCRYPT_ALG_HANDLE s_aes_ctr_algorithm_handle = NULL;
+static BCRYPT_ALG_HANDLE s_aes_keywrap_algorithm_handle = NULL;
+
+struct aes_bcrypt_cipher {
+    struct aws_symmetric_cipher cipher;
+    BCRYPT_ALG_HANDLE alg_handle;
+    /* the loaded key handle. */
+    BCRYPT_KEY_HANDLE key_handle;
+    /* Used for GCM mode to store IV, tag, and aad */
+    BCRYPT_AUTHENTICATED_CIPHER_MODE_INFO *auth_info_ptr;
+    /* Updated on the fly for things like constant-time CBC padding and GCM hash chaining */
+    DWORD cipher_flags;
+    /* For things to work, they have to be in 16 byte chunks in several scenarios. Use this
+       Buffer for storing excess bytes until we have 16 bytes to operate on. */
+    struct aws_byte_buf overflow;
+    /* This gets updated as the algorithms run so it isn't the original IV. That's why its separate */
+    struct aws_byte_buf working_iv;
+    /* A buffer to keep around for the GMAC for GCM. */
+    struct aws_byte_buf working_mac_buffer;
+};
+
+static void s_load_alg_handles(void *user_data) {
+    (void)user_data;
+
+    /* this function is incredibly slow, LET IT LEAK*/
+    NTSTATUS status = BCryptOpenAlgorithmProvider(&s_aes_cbc_algorithm_handle, BCRYPT_AES_ALGORITHM, NULL, 0)
+        AWS_ASSERT(s_aes_algorithm_handle && "BCryptOpenAlgorithmProvider() failed");
+
+    status = BCryptSetProperty(
+        s_aes_cbc_algorithm_handle,
+        BCRYPT_CHAINING_MODE,
+        (PUCHAR)BCRYPT_CHAIN_MODE_CBC,
+        (ULONG)(wcslen(BCRYPT_CHAIN_MODE_CBC) + 1),
+        0);
+
+    AWS_FATAL_ASSERT(NT_SUCCESS(status) && "BCryptSetProperty for CBC chaining mode failed");
+
+    /* Set up GCM algorithm */
+    status = BCryptOpenAlgorithmProvider(&s_aes_gcm_algorithm_handle, BCRYPT_AES_ALGORITHM, NULL, 0);
+    AWS_ASSERT(s_aes_gcm_algorithm_handle && "BCryptOpenAlgorithmProvider() failed");
+
+    status = BCryptSetProperty(
+        s_aes_gcm_algorithm_handle,
+        BCRYPT_CHAINING_MODE,
+        (PUCHAR)BCRYPT_CHAIN_MODE_GCM,
+        (ULONG)(wcslen(BCRYPT_CHAIN_MODE_GCM) + 1),
+        0);
+
+    AWS_FATAL_ASSERT(NT_SUCCESS(status) && "BCryptSetProperty for GCM chaining mode failed");
```

Comment:
trivial: why the mix of ASSERT() and FATAL_ASSERT() throughout this function?
seems fine to always FATAL_ASSERT, since this function only ever runs once